### PR TITLE
Add a class for partitions of intervals

### LINKED ISSUE #907: Create a clean distributed-ginkgo branch
To ease the development of distributed functionality for ginkgo, there should exist a branch `distributed-ginkgo` that contains the current and agreed-upon status.

For this, I want to incrementally add the functionality that is already in the branch `mpi-base-dist-mat` into the new branch. Currently, I think this will require PR for the following functionality:
- [x] MPI bindings #908  (1) 
- [x] a partition of global indices #909  (2)
- [x] distributed vectors (3) #961 
- [x] distributed matrices (4) #971
- [x] distributed solvers (5) #976 
- [ ] enable logging of MPI communications (6)
- [ ] add distributed domain decomposition predonditioner (7)
- [ ] figure out mutable access to local vectors and matrices (8) #1159 

There are the following dependencies between these tasks:
- (5) → (4) & (3)
- (4) & (3) → (2) & (1)
- (6) → (1)
- (7) → (4) & (3)
- (8) → (4) 

At the moment, there are no dependencies between (1) and (2).

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR adds a partition class, which handles an interval that is distributed across multiple ranks. 

Partially closes #907.

Main contributions are from @upsj.

### CODE REVIEW

**Comment by Slaedr** [comment]
@upsj Okay. In that case, where will the Partition class be used? Only for setting up the initial communication patterns for SpMVs etc.? Can the Partition object be deleted after that?

**Comment by upsj** [comment]
@Slaedr yes, in my [original design](https://github.com/ginkgo-project/ginkgo/blob/distributed-matrix/include/ginkgo/core/distributed/matrix.hpp) it is only used inside read_distributed, but I think @MarcelKoch needed to modify it a bit to make repartitioning work?

File: cuda/distributed/partition_kernels.cu
Comment (Slaedr) [inline cuda/distributed/partition_kernels.cu:57]:
```suggestion
                            const comm_index_type* range_parts, size_type num_ranges,
```

File: cuda/distributed/partition_kernels.cu
Comment (Slaedr) [inline cuda/distributed/partition_kernels.cu:59]: To make it clearer:
```suggestion
                            LocalIndexType* local_start_indices, LocalIndexType* part_sizes)
```

File: cuda/distributed/partition_kernels.cu
Comment (Slaedr) [inline cuda/distributed/partition_kernels.cu:100]: Again, it would be nice to have the formal parameters with the same names as the arrays in the surrounding code.

File: dpcpp/distributed/partition_kernels.dp.cpp
Comment (Slaedr) [inline dpcpp/distributed/partition_kernels.dp.cpp:84]: Since this code is not particularly trivial, I suggest you remove this kernel and put it in a common header file; perhaps `common/unified/partition_kernels.hpp`? Or any other place you think is appropriate. This could apply to the kernel above as well.

File: dpcpp/distributed/partition_kernels.dp.cpp
Comment (Slaedr) [inline dpcpp/distributed/partition_kernels.dp.cpp:48]: Since this kernel is not yet supposed to work.
```suggestion
{
    GKO_NOT_IMPLEMENTED;
```

File: hip/distributed/partition_kernels.hip.cpp
Comment (Slaedr) [reply hip/distributed/partition_kernels.hip.cpp:61]: Instead, I would suggest moving the common kernels below to `common/unified` so that they can be used in DPC++ as well. The rest (thrust calls etc.) could be left here.

File: omp/distributed/partition_kernels.cpp
Comment (Slaedr) [inline omp/distributed/partition_kernels.cpp:60]: This kernel could be replaced by the same code as for cuda/hip. Thrust can, in fact, be used on the host with an openmp backend as well. You'll just have to look carefully at what Thrust macro to define where so that it uses OpenMP in this file.

I guess one issue is we don't look for Thrust when we're not building the cuda or hip backends. If we don't want to add a thrust dependency for that, perhaps it makes sense to keep this.

File: test/distributed/partition_kernels.cpp
Comment (Slaedr) [inline test/distributed/partition_kernels.cpp:288]:
```suggestion
TYPED_TEST(Partition, BuildsFromContiguousWithMostlyEmptyParts)
```

**Review by Slaedr — COMMENTED** [review]
Great work on the algorithms. I guess there are a couple of things left:
- core tests
- Openmp tests if we decide to keep the separate implementation.

Apart from that, I have some comments below.

File: common/unified/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp:107]: The PR #676 has a safer sentinel value, which does basically the same, so we will use that one after that PR is merged. I don't think it is necessary to add an assert until then.

File: common/unified/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp:75]: All array of the partition class are filled with 0, so `ranges[0]` doesn't need to be changed. I will check if setting `bounds[0]` is even necessary.

Comment (upsj) [reply common/unified/distributed/partition_kernels.cpp:75]: That is mainly a simplification like row pointers, where we could theoretically omit the initial 0, but it makes implementations much easier at little overhead.

File: common/unified/distributed/partition_kernels.cpp
Comment (Slaedr) [reply common/unified/distributed/partition_kernels.cpp:107]: What is the safer sentinel value?

Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp:107]: See https://github.com/ginkgo-project/ginkgo/blob/e18f33a663b3fe29263c4361e02553b6adc82432/include/ginkgo/core/base/types.hpp#L637
It basically does what you suggest.

File: omp/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply omp/distributed/partition_kernels.cpp:60]: Yes, I don't think the benefit from unifying it is bigger than the hassle from using/not using thrust, so I will keep it as it is.

File: test/distributed/partition_kernels.cpp
Comment (yhmtsai) [inline test/distributed/partition_kernels.cpp:449]: can also use generate_random_array

File: common/unified/distributed/partition_kernels.cpp
Comment (yhmtsai) [inline common/unified/distributed/partition_kernels.cpp:78]: range_bounds[0]?

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (yhmtsai) [inline common/unified/distributed/partition_kernels.hpp.inc:81]: Is it due to sentinel? I thought it due to the size and index?

**Review by yhmtsai — CHANGES_REQUESTED** [review]
the public interface in partition.hpp needs `@return` or `@param`
common/unified/distributed/partition_kernels.hpp.inc should be in cuda_hip if it only be used for cuda_hip.
`common/cuda_hip/distributed/partition_kernels.hpp.inc` should be in cuda and hip separately if we stay the original design.

File: common/unified/distributed/partition_kernels.cpp
Comment (Slaedr) [reply common/unified/distributed/partition_kernels.cpp:107]: Now I guess we can use `invalid_index` here?

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (Slaedr) [inline common/unified/distributed/partition_kernels.hpp.inc:33]: If you want to keep the guard, maybe
```suggestion
#ifndef GKO_COMMON_UNIFIED_PARTITION_KERNELS_HPP
```

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (Slaedr) [inline common/unified/distributed/partition_kernels.hpp.inc:79]:
```suggestion
                i > 0 ? grouped_part_ids[i - 1] : invalid_index<comm_index_type>();
```

File: common/unified/distributed/partition_kernels.cpp
Comment (Slaedr) [inline common/unified/distributed/partition_kernels.cpp:107]:
```suggestion
            const auto prev_part = i > 0 ? mapping[i - 1] : invalid_index<comm_index_type>();
```

File: core/distributed/partition.cpp
Comment (Slaedr) [reply core/distributed/partition.cpp:88]: It might be a good idea to have a check only in debug builds.

File: core/distributed/partition.cpp
Comment (Slaedr) [reply core/distributed/partition.cpp:84]: If range `i` contains one element, say 3, it would be [3,4), so `ranges[i] == 3` and `ranges[i+1] == 4`. So I guess, indeed, a valid `ranges` must have at least two entries.

Comment (MarcelKoch) [reply core/distributed/partition.cpp:84]: Sorry about this check, that is some WIP that I forgot to leave out. I will remove that again, and add some validation support when that becomes possible.

File: common/unified/distributed/partition_kernels.cpp
Comment (Slaedr) [reply common/unified/distributed/partition_kernels.cpp:160]: I think `&&` might be better here, because we really mean a full AND here, not a bitwise AND. The bitwise AND should also work, but then we are relying on the cast to always put the digit 1 in the same position while casting from `true`. I'm not sure what the standard says, though that's probably always the case. But still, I doubt it matters for performance and `&&` is clearer here.

File: common/unified/distributed/partition_kernels.cpp
Comment (Slaedr) [reply common/unified/distributed/partition_kernels.cpp:78]: I think that's taken care of because it's initialized to 0 in the constructor. But, for robustness w.r.t. future changes in the constructor, it might be better to explicitly set it here.

**Review by Slaedr — APPROVED** [review]
I agree with some of @yhmtsai's points. Once those are addressed, I think this is good to go. 

(I guess we're all okay with storing the entire partition everywhere. 😅 )

File: core/distributed/partition.cpp
Comment (MarcelKoch) [reply core/distributed/partition.cpp:88]: At the moment it just garbage in, garbage out. A check could be added as part of the validation overhaul.

File: omp/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply omp/distributed/partition_kernels.cpp:87]: I don't think that is necessary, since we don't use the `nowait` clause

File: reference/test/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply reference/test/distributed/partition_kernels.cpp:221]: I think it's necessary for our AAA pattern convention.

File: omp/distributed/partition_kernels.cpp
Comment (upsj) [reply omp/distributed/partition_kernels.cpp:87]: Might make sense to double-check that, I've been bitten by such an assumption before (which is also why we have a barrier above)

File: common/unified/distributed/partition_kernels.cpp
Comment (yhmtsai) [reply common/unified/distributed/partition_kernels.cpp:160]: I guess you mean from logical point?
each entry is true/false -> 1/0 and we reduce them here. They are same meaning if they always 1/0.
If to keep the logical point, I would suggest we have explicit cast here.

File: omp/distributed/partition_kernels.cpp
Comment (yhmtsai) [reply omp/distributed/partition_kernels.cpp:87]: @MarcelKoch Could you point me where I can get the `omp for` has sync in the end?
the previous for is normal for in each thread not `#omp for` so it needs the sync, I think. Or the `#omp for` also has a sync in the beginning

File: reference/test/distributed/partition_kernels.cpp
Comment (yhmtsai) [reply reference/test/distributed/partition_kernels.cpp:221]: I see, it's also okay.
we do not consider the `using...` in any AAA section, so it always go to the nearest AAA section.
If you consider it is belong to Assign section, I think it should make sense, too.

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (yhmtsai) [reply common/unified/distributed/partition_kernels.hpp.inc:33]: GKO_COMMON... is more proper.
the format_header only consider the first guard with `#ifndef GKO_...`, so it does not catch this and format it.

File: omp/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply omp/distributed/partition_kernels.cpp:87]: I've looked at the 4.5 standard specification https://www.openmp.org/wp-content/uploads/openmp-4.5.pdf, and I interpret the section on p.205 l.17-22 on the `reduction` clause as if it will synchronize the result at the end of the reduction.
But perhaps that is not true if we rely on an older standard.

File: test/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply test/distributed/partition_kernels.cpp:449]: thanks, didn't know about that function.

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.hpp.inc:33]: I've removed the guard, since we don't use it in these cases.

File: common/unified/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp:78]: I set it now explicitly in the kernel.

File: common/unified/distributed/partition_kernels.hpp.inc
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.hpp.inc:77]: The sentinel value is set in the kernel above, so it is correct, but also confusing, since I've extracted both kernels into their own functions. I've adjusted the comment to reflect that.

File: common/unified/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp:166]: In which direction do mean the cast? For me, the only reasonable is casting the result to `uint32`. Also, I would like to stay with `&&` since it's clearer, as @Slaedr mentioned.

Comment (upsj) [reply common/unified/distributed/partition_kernels.cpp:166]: Even though it's not explicitly specified (I will take care of that though!), run_kernel_reduction should internally assign every result to the ValueType deduced for the reduction, i.e. the identity element parameter. So we don't need to worry about returning a different type here.

Comment (MarcelKoch) [reply common/unified/distributed/partition_kernels.cpp]: I've added the cast to `unit32`, although I don't think it should matter too much.

File: omp/distributed/partition_kernels.cpp
Comment (yhmtsai) [reply omp/distributed/partition_kernels.cpp:91]: from 3.1, https://www.openmp.org/wp-content/uploads/OpenMP3.1.pdf
p105 l.13 - 19 gives the similar behavior

**Comment by yhmtsai** [comment]
LGTM for kernel side. but the unified stuff looks weird to me.
`common/unified/distributed/partition_kernels.hpp.inc` is used in `common/cuda_hip/distributed/partition_kernels.hpp.inc`

`common/cuda_hip/distributed/partition_kernels.hpp.inc` should avoid including the file because we do not have guard there.
unified is supposed to work on all executor except for reference, so removing dpcpp from test/kernel is weird to me.
which part of the unified code doesn't work on dpcpp? they should work on dpcpp, I think

**Comment by MarcelKoch** [comment]
@yhmtsai The work for dpcpp was not finished and it's not a priority for me. At some point we discussed, that we should rather proceed without the dpcpp kernels than wait until that is fixed. But in the end, it would also use the kernels from `common/unified/distributed/partition_kernels.hpp.inc`, that's why I left it there instead of putting it into the `cuda_hip` file. I agree that using a `*.hpp.inc` file in `common/unified` looks a bit weird, but since the usage is more similar to the `cuda_hip` files, I went this way.

I don't think an include guard is necessary, since `partition_kernel.hpp.inc` won't be included twice somewhere, and leaving out the include guard is more consistent with the rest of the `common` files.

Regarding the test, the alternative would be to write the same tests three times, which seems unnecessary to me. Also, the dpcpp will become eventually available, so I think is justified to have the common tests, with the one exception.

**Comment by yhmtsai** [comment]
Yes, we do not put the guard in `cuda_hip/*.inc` file. I am not saying the guard is needed when we use cuda_hip like a macro way.
We do not include any inc file in inc and it should explicitly be include from the file such that we know the file are copied to here. include inc in inc will make us ignore that. we may include two inc and they have the duplicated item.
That's a reason why I think we should avoid include a file inside inc when we do not have guard there.

if the unified/*.inc if only used in cuda_hip, can it not be in cuda_hip?
I agree with processing the pr without dpcpp support, but it might be considered moved out of unified.
If I understand it correctly, the kernels in unified should also work on dpcpp. you can test the kernel directly.

test can test them the directly not from the constructor so you will not face the build_start_indices not implemented in dpcpp. we also have the duplicated test in hip/cuda/omp actually.
Using disable_executor makes sense in duplication aspect although it changes my first thought on the test folder.
I draw back my comment on moving test. the changes on duplication is more painful.

Could you also add the TODO inside the code? for example the num_range == 1 check situation, or something else is mentioned in the review.

**Comment by upsj** [comment]
The code will be used in DPC++ as well once https://github.com/oneapi-src/oneDPL/pull/388 is merged and released. This issue is the only reason we can't merge the DPC++ code right away.

**Comment by MarcelKoch** [comment]
I will put the kernels from `partition_kernels.hpp.inc` into the `common/cuda_hip` file directly. We can refactor that again, when the dpcpp stuff becomes ready. I hope that will resolve all these issues.

I will add a todo for the dpcpp implementation, but I think the num_range == 1 check I had previously is now addressed in the documentation.

However, I will wait on pushing these changes, until I've fixed the cuda 9.2 issue.

File: reference/test/distributed/partition_kernels.cpp
Comment (yhmtsai) [inline reference/test/distributed/partition_kernels.cpp:166]: what happen when it is not zero?

File: include/ginkgo/core/distributed/partition.hpp
Comment (yhmtsai) [inline include/ginkgo/core/distributed/partition.hpp:255]: I guess here? I can not get the part how to handle one element. Do I miss anything?

File: test/distributed/partition_kernels.cpp
Comment (yhmtsai) [inline test/distributed/partition_kernels.cpp:142]: because use many gko::distributed::Partition<local_index_type, global_index_type>, maybe use alias for it

File: include/ginkgo/core/distributed/partition.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/distributed/partition.hpp:255]: You're right, I was thinking about something else. To clarify, I added the comment to prevent issues from using `size(ranges) - 1`. If you pass just one element you will still get an empty range.

File: reference/test/distributed/partition_kernels.cpp
Comment (MarcelKoch) [reply reference/test/distributed/partition_kernels.cpp:166]: I've clarified the documentation, that the first element of the ranges needs always to be 0. Still, it wouldn't matter, you would still get an empty partition, but with `range_bounds = {0}`, regardless of what element was passed.

File: include/ginkgo/core/distributed/partition.hpp
Comment (yhmtsai) [reply include/ginkgo/core/distributed/partition.hpp:255]: I see. that makes sense

File: reference/test/distributed/partition_kernels.cpp
Comment (yhmtsai) [reply reference/test/distributed/partition_kernels.cpp:166]: Thanks for clarification

**Review by yhmtsai — APPROVED** [review]
LGTM. Is the cuda9.2 issue also from executing some kernel with zero grid size?

**Comment by upsj** [comment]
@yhmtsai Yes, as far as I can tell, Thrust internally checks whether a previous kernel launch failed, and even though this kind of failure never cascades, it means that any Thrust kernel call after an empty CUDA kernel call causes an exception to be thrown.

**Comment by yhmtsai** [comment]
I see. There were some kernels with zero grid size but do not give the error in gtest because no thrust kernel after that to raise it.

**Pull request merged**
